materialize-rockset/rockset.go

```diff
@@ -67,31 +77,98 @@ func getCollection(ctx context.Context, client *client, workspace string, collec
-			} else if col.Status != "READY" {
-				// Not ready yet, so wait a couple of seconds then we'll try again.
-				time.Sleep(time.Second * 2)
-			} else {
-				return col, nil
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		var collection, err = getCollection(ctx, client, workspace, collectionName)
+		if err != nil {
+			return err
+		}
+		if collection == nil {
+			return fmt.Errorf("rockset collection '%s' does not exist or has been deleted", collectionName)
+		}
+
+		var ready = collection.Status == STATUS_READY
+
+		// We'll check to see if a collection has made no progress and emit a specific warning in that case.  If a
+		// Rockset collection isn't using a large enough instance, then it'll silently refuse to actually use the
+		// integration that was configured. This is a really easy mistake to make, and there doesn't appear to be a way
+		// to actually detect the instance size via the API, so we'll try to log a helpful warning if the collection
+		// goes too long without downloading any of the objects from cloud storage.
+		var bulkIngestStalled = false
+		var collectionAge = time.Since(collection.CreatedAt)
+		// If an integration was specified, then we need to _also_ wait for that integration to fully process all of its
+		// files. Technically, this will already be the case if there was enough data in the source bucket to trigger a
+		// "bulk load", because in that case Rockset will not set the status to READY until after all the files have
+		// been uploaded. BUT, if the source bucket didn't contain enough data to trigger the "bulk load", then the
+		// status may be set to READY before all of the source files have been processed. This condition is a guard
+		// against ingesting data out of order in that specific case.
+		if integration != "" {
+			var source = collection.GetCloudStorageSource(integration)
+			if source == nil {
+				return fmt.Errorf("expected collection '%s' to have a compatible cloud-storage integration named '%s', but no such source exists", collectionName, integration)
 			}
+			ready = ready && source.ObjectCountTotal == source.ObjectCountDownloaded
+			bulkIngestStalled = source.ObjectCountTotal > 0 && source.ObjectCountDownloaded == 0 && collectionAge > time.Minute*5
 		}
+
+		var logEntry = log.WithFields(log.Fields{
+			"rocksetWorkspace":  workspace,
+			"rocksetCollection": collection,
+		})
+		if ready {
+			logEntry.Info("Rockset collection has completed bulk ingestion")
+			return nil
+		} else if bulkIngestStalled {
+			logEntry.Warn("Rockset collection integration has not made any progress. Is your Rockset virtual instance large enough to support bulk ingestion?")
+		} else {
+			logEntry.Info("Rockset collection integration has not yet caught up (will retry)")
+		}
+
+		time.Sleep(nextIngestCompletionBackoff(collectionAge))
 	}
 }
+
+func nextIngestCompletionBackoff(timeSinceCreation time.Duration) time.Duration {
```

Comment:
We need to have 5gb to even bother with the bulk ingestion mode, right? Do you have any ballpark of how long that might take to complete? Should we even check in the first 5 minutes?

Reply:
Technically, we await completion at the start of every processing term, so I think it's best to just check early so we don't hold things up in the common case where the collection is already `READY`. Additionally, we create the collection during `Apply` and do this await during `Transactions`, so there could already be some time that's passed.